### ISSUE TITLE
OMP upgrade to 3.4.0

### ISSUE THREAD

**Issue by tjakopec** [pkp/pkp-lib#9186]
**Describe the bug**
Upgrade script fail on https://github.com/pkp/omp/blob/38048f9b30506296ca2d47a109c69a6ffed7bed2/classes/migration/upgrade/v3_4_0/I6782_MetricsSubmission.php#L52

database (10.4.21-MariaDB) log is as follows

insert into `metrics_submission` (`load_id`, `context_id`, `submission_id`, `chapter_id`, `representation_id`, `submission_file_id`, `file_type`, `assoc_type`, `date`, `metric`) select m.load_id, m.context_id, m.submission_id, CAST(sfs.setting_value AS UNSIGNED), m.representation_id, m.assoc_id, m.file_type, m.assoc_type, DATE_FORMAT(STR_TO_DATE(m.day, '%Y%m%d'), '%Y-%m-%d'), m.metric from `metrics` as `m` left join `submission_file_settings` as `sfs` on `m`.`assoc_id` = `sfs`.`submission_file_id` and `m`.`assoc_type` = ? and `sfs`.`setting_name` = ? where `m`.`assoc_type` = ? and `m`.`metric_type` = ?

insert into `metrics_submission` (`load_id`, `context_id`, `submission_id`, `chapter_id`, `representation_id`, `submission_file_id`, `file_type`, `assoc_type`, `date`, `metric`) select m.load_id, m.context_id, m.submission_id, CAST(sfs.setting_value AS UNSIGNED), m.representation_id, m.assoc_id, m.file_type, m.assoc_type, DATE_FORMAT(STR_TO_DATE(m.day, '%Y%m%d'), '%Y-%m-%d'), m.metric from `metrics` as `m` left join `submission_file_settings` as `sfs` on `m`.`assoc_id` = `sfs`.`submission_file_id` and `m`.`assoc_type` = 515 and `sfs`.`setting_name` = chapterId where `m`.`assoc_type` = 515 and `m`.`metric_type` = omp::counter

**Edit: This is a red herring; see below for error message**
~~`sfs`.`setting_name` = chapterId    -> without single quotation mark~~
~~`m`.`metric_type` = omp::counter -> without single quotation mark~~

~~if I change line https://github.com/pkp/omp/blob/38048f9b30506296ca2d47a109c69a6ffed7bed2/classes/migration/upgrade/v3_4_0/I6782_MetricsSubmission.php#L52 to~~

~~$join->where('sfs.setting_name', '=', '\'chapterId\'');~~

~~then in database log i get `sfs`.`setting_name` = '\'chapterId'\'~~

**To Reproduce**
Upgrade from OMP 3.3.0.14

**What application are you using?**
OMP 3.3.0.14

**Additional information**

**Comment by asmecher** [comment]
@tjakopec, I think this might be a side-effect of how the database query is logged, rather than an actually missing `'` string delimiter. Can you provide the error message that you got when the upgrade failed?

**Comment by tjakopec** [comment]
@asmecher here is error message

2023-07-26 05:09:45 [migration: APP\migration\upgrade\v3_4_0\I6782_MetricsSeries]
2023-07-26 05:09:45 [migration: APP\migration\upgrade\v3_4_0\I6782_MetricsSubmission]
2023-07-26 05:09:46 [revert migration: APP\migration\upgrade\v3_4_0\I6782_MetricsSeries]
2023-07-26 05:09:46 [downgrade for "APP\migration\upgrade\v3_4_0\I6782_MetricsSeries" unsupported: Downgrade not supported]
ERROR: Upgrade failed: DB: SQLSTATE[23000]: Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails (`omp_prod`.`metrics_submission`, CONSTRAINT `metrics_submission_chapter_id_foreign` FOREIGN KEY (`chapter_id`) REFERENCES `submission_chapters` (`chapter_id`) ON DELETE CASCADE) (SQL: insert into `metrics_submission` (`load_id`, `context_id`, `submission_id`, `chapter_id`, `representation_id`, `submission_file_id`, `file_type`, `assoc_type`, `date`, `metric`) select m.load_id, m.context_id, m.submission_id, CAST(sfs.setting_value AS UNSIGNED), m.representation_id, m.assoc_id, m.file_type, m.assoc_type, DATE_FORMAT(STR_TO_DATE(m.day, '%Y%m%d'), '%Y-%m-%d'), m.metric from `metrics` as `m` left join `submission_file_settings` as `sfs` on `m`.`assoc_id` = `sfs`.`submission_file_id` and `m`.`assoc_type` = 515 and `sfs`.`setting_name` = chapterId where `m`.`assoc_type` = 515 and `m`.`metric_type` = omp::counter)

**Comment by asmecher** [comment]
@bozana, can you have a look at this one? The issue problem is described [in this comment](https://github.com/pkp/pkp-lib/issues/9186#issuecomment-1650990125) -- the original report about the string quoting is a red herring.

**Comment by bozana** [comment]
Hi @tjakopec, could you please execute this query in your DB:
```
select sfs.setting_value from submission_file_settings as sfs 
left join submission_chapters sc on sfs.setting_value = sc.chapter_id
where setting_name = 'chapterId' AND sfs.setting_value IS NOT NULL AND sc.chapter_id IS NULL
```
I suppose that there is a `chapterId` in your DB table `submission_file_settings` that is not in the DB table `submission_chapters`. We did not check after the orphaned chapter IDs in our upgrade preflight check :-( If that is the reason I would add that check...
Else, I will investigate further what else could be the reason...

**Comment by tjakopec** [comment]
Hi @bozana,
 here is result of query

```
select sfs.setting_value from submission_file_settings as sfs 
left join submission_chapters sc on sfs.setting_value = sc.chapter_id
where setting_name = 'chapterId' AND sfs.setting_value IS NOT NULL AND sc.chapter_id IS NULL
```

365
366
367
368
369
370
371
372
373
374
375
376
377
378
406
379
380
381
382
383
384
385
386
387
388
389
390
391
392
393
394
395
396
397
398
399
400
401
402
403
404
405

**Comment by bozana** [comment]
Hmmm... So yes, it seems that for some reason those chapter IDs are in your DB table submission_file_settings but not in the table submission_chapters. Maybe you can investigate how come? Are the files having those chepter IDs there?
Hmmm... OK... I will consider this case (orphaned chapter IDs) in the upgrade check...

**Comment by tjakopec** [comment]
Hi @bozana,
 is it possible that those submission_chapters where deleted because publications id in those submission_chapters table where deleted - thats for sure.

For example in manageCatalog pard of admin interface there is monograph id 56 and 58, missing 57. And that 57 have current_publication_id is under submission id 67 that have file with id 1170 that have settings_value 366 that is missing in submission_chapters table.

Hope this helps. If you want I can send you DB dump in private email for inspection. My mail is tjakopec@gmail.com and feel free to contact me.

**Comment by bozana** [comment]
Hi @tjakopec, thanks a lot! Yes, that could be...
I'll contact you regarding DB dump... -- to be sure how to fix the problem...

**Comment by bozana** [comment]
After analyses we figured out that the problem exists only in one submission. It seems like some chapter deletions did not work fully there -- the related entries from `submission_file_settings` were not deleted.
The files (with the not existing chapter IDs) do exist in the system, are listed under Publication > Publication Formats and approved/available. Because connected to the chapter IDs that do not exist, the files are not listed on the monograph view page. The files seem to be a copy of other published files that are correctly connected to the chapters. Thus, it seems like editors tried to delete chapters (and for some reason it did not work correctly), uploaded the files anew, created new chapters, and connected the new files correctly to new chapters.
If we would only remove the entries containing the not existing chapter IDs from the table `submission_file_settings`, the files would be listed on the monograph view page, as monograph (and not chapter) files.
Thus, we decided that editors should manually fix this case.
